f3/granite.go

```diff
@@ -246,14 +250,21 @@ func (i *instance) receiveOne(msg *GMessage) {
 		round.prepared.Receive(msg.Sender, msg.Value)
 	case COMMIT:
 		round.committed.Receive(msg.Sender, msg.Value)
+	case DECIDE:
+		i.decision.Receive(msg.Sender, msg.Value)
 	default:
 		i.log("unexpected message %v", msg)
 	}
 
+	// Bracha amplification for DECIDE phase (instead of justifying DECIDE messages)
+	if msg.Step == DECIDE {
+		i.maybeAmplify(msg.Value)
+	}
+
 	// Try to complete the current phase.
 	// Every COMMIT phase stays open to new messages even after the protocol moves on to
 	// a new round. Late-arriving COMMITS can still (must) cause a local decision, *in that round*.
-	if msg.Step == COMMIT {
+	if msg.Step == COMMIT && i.phase != TERMINATED {
```

Comment:
```suggestion
	if msg.Step == COMMIT && i.phase != DECIDE {
```

Please move the `TERMINATED` check to the top of `receiveOne`, and just return as a no-op. This is possible if the inbox has more than one message in it, and the first one results in a termination. Otherwise, receiving a new message while TERMINATED already results in a rejection.

But here we still need a `!= DECIDE` check. If this node has entered DECIDE phase because it has observed a strong quorum of COMMIT, it can't make a phase change based on more of them so we should not process it.